### ISSUE TITLE
Consider Rust BPE tokenizer implementation

### ISSUE THREAD

**Issue by tarekgh** [dotnet/machinelearning#7336]
Try to look at rust gems BPE tokenizer performance and figure out if it is worth adapting it in the ML tokenizers library. 

https://github.com/github/rust-gems/blob/main/crates/bpe/README.md

**Comment by rosebyte** [comment]
Triage: is it possible that this issue was actually addressed by https://github.com/dotnet/machinelearning/pull/7580? @tarekgh, do you please remember?

**Comment by tarekgh** [comment]
No, I don't think #7580 addresses this one. #7580 added a heap-based fallback for large inputs, which is tiktoken's own `_byte_pair_merge_large`. The rust-gems README actually calls that heap approach the O(n log n) "baseline," not its real contribution, so #7580 is closer to the thing rust-gems compares against than to rust-gems itself.

The point of this issue is the rust-gems feature set, which is aimed at chunking rather than plain encoding speed: linear worst-case encoding, token counting for sub-ranges, correct incremental counting while appending text, and splitting after exactly n tokens at a character boundary. #7580 does not provide any of those.

So the perf motivation is partly reduced now that we handle the large-input worst case, but the actual evaluation asked for here is still open. Unless we have a concrete scenario that needs the sub-range or incremental counting and chunking APIs, I'd lean toward closing this as not planned, since adopting a new algorithm is a significant cost without a consumer asking for those capabilities. If someone has such a scenario, let's capture it here and keep it open.
